easybuild/main.py

```diff
@@ -213,162 +214,109 @@ def clean_exit(logfile, tmpdir, testing, silent=False):
     sys.exit(0)
 
 
-def main(args=None, logfile=None, do_build=None, testing=False, modtool=None):
+def process_easystack(easystack_path, args, logfile, testing, init_session_state, do_build):
     """
-    Main function: parse command line options, and act accordingly.
-    :param args: command line arguments to use
+    Process an EasyStack file. That means, parsing, looping over all items in the EasyStack file
+    building (where requested) the individual items, etc
+
+    :param easystack_path: path to EasyStack file to be processed
+    :param args: original command line arguments as received by main()
     :param logfile: log file to use
-    :param do_build: whether or not to actually perform the build
     :param testing: enable testing mode
+    :param init_session_state: initial session state, to use in test reports
+    :param do_build: whether or not to actually perform the build
     """
-
-    register_lock_cleanup_signal_handlers()
-
-    # if $CDPATH is set, unset it, it'll only cause trouble...
-    # see https://github.com/easybuilders/easybuild-framework/issues/2944
-    if 'CDPATH' in os.environ:
-        del os.environ['CDPATH']
-
-    # When EB is run via `exec` the special bash variable $_ is not set
-    # So emulate this here to allow (module) scripts depending on that to work
-    if '_' not in os.environ:
-        os.environ['_'] = sys.executable
-
-    # purposely session state very early, to avoid modules loaded by EasyBuild meddling in
-    init_session_state = session_state()
-
-    eb_go, cfg_settings = set_up_configuration(args=args, logfile=logfile, testing=testing)
-    options, orig_paths = eb_go.options, eb_go.args
+    orig_paths, opts_per_ec = parse_easystack(easystack_path)
 
     global _log
-    (build_specs, _log, logfile, robot_path, search_query, eb_tmpdir, try_to_generate,
-     from_pr_list, tweaked_ecs_paths) = cfg_settings
-
-    # load hook implementations (if any)
-    hooks = load_hooks(options.hooks)
-
-    run_hook(START, hooks)
-
-    if modtool is None:
+    _log.debug("Start build loop over items in the EasyStack file: %s" % orig_paths)
+    # TODO: insert fast loop that validates if all command line options are valid. If there are errors in options,
+    # we want to know early on, and this loop potentially builds a lot of packages and could take very long
+    # for path in orig_paths:
+    #     validate_command_opts(args, opts_per_ec[path])
+
+    # Loop over each item in the EasyStack file, each time updating the config
+    # This is because each item in an EasyStack file can have options associated with it
+    do_cleanup = True
+    for path in orig_paths:
```

Comment:
This should become a for loop over tuples, so we can support the use case of having the same easyconfig file listed multiple times in an easystack file, with different options (with and without `--from-pr`, for example):

```python
easystack = parse_easystack(easystack_path)
...
for (path, opts) in easystack:
    ...
    if opts:
        ...
        ec_args = opts_dict_to_eb_opts[opts]
```

Reply:
Done in https://github.com/easybuilders/easybuild-framework/pull/4057/commits/7e2e6d60bc028033c6fe587b49901781e994950c